modules/ximgproc/src/fbs_filter.cpp

```diff
@@ -487,12 +488,18 @@ namespace ximgproc
         cv::Mat filtered_disp;
         float fgs_colorSigma = 1.5;
         float fgs_spatialSigma = 2000;
-	      target.convertTo(x, CV_32FC1, 1.0f/255.0f);
-	      confidence.convertTo(w, CV_32FC1);
+
+        if(target.depth() == CV_16U)
+            target.convertTo(x, CV_32FC1, 1.0f/65535.0f);
+        else if(target.depth() == CV_8U)
+            target.convertTo(x, CV_32FC1, 1.0f/255.0f);
+
+        confidence.convertTo(w, CV_32FC1);
+
         xw = x.mul(w);
         cv::ximgproc::fastGlobalSmootherFilter(guide, xw, filtered_xw, fgs_spatialSigma, fgs_colorSigma);
         cv::ximgproc::fastGlobalSmootherFilter(guide, w, filtered_w, fgs_spatialSigma, fgs_colorSigma);
-        cv::divide(filtered_xw, filtered_w, filtered_disp, 1.0f, CV_32FC1);
+        cv::divide(filtered_xw, filtered_w+EPS, filtered_disp, 1.0f, CV_32FC1);
```

Comment:
Please add some simple regression test to for this issue into test_fbs.cpp file.
New test images can be added into opencv_extra/testdata (but prefer to reuse existed).

Is it related for master branch only? Or 3.4 too? OpenCV handles division by zero differently for floating-point numbers (3.4 has zero pre-check, master uses IEEE 754 rules).

Reply:
This was related to master branch only as I was originally working with 3.4 and noticed the issue while switching to master. Like you pointed out, I had already a suspicion that the 3.4 handles division by zero differently but didn't check this.

Would it be better to remove the Fast Global Filtering because it is not used in the original paper for initializing the Fast Bilateral Solver? In fbs_filter.cpp, the parameters for FGS are also fixed. I would prefer to remove the FGS part and maybe add it to disparity filtering demo for comparison.

In the case, you think its better to keep the FGS in fbs_filter.cpp, I can add the requested test.

Reply:
> it is not used in the original paper

It make sense to remove that.
From user/applications side it is more easy to add pre-filter than to remove extra stage.

@berak Do you have any objections?

Reply:
BTW, This code is not a part of any OpenCV release yet - you can work on 3.4 branch (3.4.4 expected on this week).

Reply:
no objections, wise deccision ;)

Reply:
Alright, thanks! I will try to revise the code as soon as possible.

Reply:
In its present state, the fbs filter expects float confidence map values to be in [0,255] range like in disparity_filters.cpp. Would it be better to change the float confidence values to [0,1] range instead?

Besides this issue my revision should be complete.

Reply:
> float confidence values to [0,1] range

Sounds good to me.

Reply:
Now the FGS (i.e. WLS) filtering is moved from fbs_filter.cpp to disparity_filtering.cpp as a preprocessing option. Consequently, there is no division by zero issue even between master and 3.4 branches.

Reply:
@alalek I added the fixed FBS filter to 3.4 branch (hopefully wasn't too late for 3.4.4).

Reply:
> Please add some simple regression test to for this issue into test_fbs.cpp file.
> New test images can be added into opencv_extra/testdata (but prefer to reuse existed).

I checked the `test_fbs.cpp` as well. It is comparing the output to a reference image produced with a different filter (FGS) with a very loose error threshold. I updated the test with a new reference image produced with the same filter and will commit it soon.